### PR TITLE
Sentinel tls support

### PR DESCRIPTION
**Pull request by MattiasLass**
I've implemented the requested changes from #2139, as the progress on the original PR seems to have stopped.

### CODE REVIEW

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 1]
```diff
@@ -30,7 +34,13 @@
...
   protected String sentinelClientName;
 
-  protected final Set<MasterListener> masterListeners = new HashSet<>();
+  protected boolean isMasterSslEnabled;
+  protected boolean isSentinelSslEnabled;
+  protected SSLSocketFactory sslSocketFactory;
+  protected SSLParameters sslParameters;
+  protected HostnameVerifier hostnameVerifier;
+
+  protected Set<MasterListener> masterListeners = new HashSet<MasterListener>();
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java:43]: Keep final keyword from earlier.
```suggestion
  protected final Set<MasterListener> masterListeners = new HashSet<>();
```

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 2]
```diff
@@ -129,14 +139,17 @@ public JedisSentinelPool(String masterName, Set<String> sentinels,
       final int sentinelConnectionTimeout, final int sentinelSoTimeout, final String sentinelPassword,
       final String sentinelClientName) {
     this(masterName, sentinels, poolConfig, connectionTimeout, soTimeout, null, password, database, clientName,
-        sentinelConnectionTimeout, sentinelSoTimeout, null, sentinelPassword, sentinelClientName);
+        sentinelConnectionTimeout, sentinelSoTimeout, null, sentinelPassword, sentinelClientName, false, false,
+        null, null, null);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java:143]: After keeping previously existing constructor, this change won't be required.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 3]
```diff
@@ -129,14 +139,17 @@ public JedisSentinelPool(String masterName, Set<String> sentinels,
...
   }
 
   public JedisSentinelPool(String masterName, Set<String> sentinels,
       final GenericObjectPoolConfig poolConfig, final int connectionTimeout, final int soTimeout,
       final String user, final String password, final int database, final String clientName,
       final int sentinelConnectionTimeout, final int sentinelSoTimeout, final String sentinelUser,
-      final String sentinelPassword, final String sentinelClientName) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java:139]: This removes an existing constructor. We can't do that.

Hint: Keep this constructor, copy previous constructor from earlier code and paste it just before this constructor.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 4]
```diff
@@ -265,7 +286,7 @@ private HostAndPort initSentinels(Set<String> sentinels, final String masterName
     return master;
   }
 
-  private HostAndPort toHostAndPort(List<String> getMasterAddrByNameResult) {
+  protected HostAndPort toHostAndPort(List<String> getMasterAddrByNameResult) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java:289]: Why change to `protected`?

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 4 above)
Comment (MattiasLass) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java]: I can't say why the original author made the change. However it does seem unnecessary so I have reverted it back to private.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 1 above)
Comment (MattiasLass) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java]: reverted the change

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 2 above)
Comment (MattiasLass) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java:143]: reverted the change

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 3 above)
Comment (MattiasLass) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java:139]: restored the missing constructor

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 4 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java]: Oh, okay.

**Review by sazzad16 — DISMISSED** [review]

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 5]
```diff
@@ -13,6 +13,10 @@
 import redis.clients.jedis.exceptions.JedisConnectionException;
 import redis.clients.jedis.exceptions.JedisException;
 
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.SSLSocketFactory;
+
 public class JedisSentinelPool extends JedisPoolAbstract {
 
   protected GenericObjectPoolConfig poolConfig;
```
Comment (bsbodden) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java:22]: `GenericObjectPoolConfig` is a raw type, how about sprinkling some `@SupressWarnings` if it cannot be parameterized with `<?>`

**Review by bsbodden — COMMENTED** [review]
Looks good, just a few minor comments. Thank you for your contribution.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 5 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java:22]: @bsbodden That's 7 years old code. No one is touching that is working. You can take the dare yourself and touch it :D

File: src/test/java/redis/clients/jedis/tests/SSLJedisSentinelPoolTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static org.junit.Assert.assertTrue;
+
+public class SSLJedisSentinelPoolTest {
+  private static final String MASTER_NAME = "mymaster";
+
+  protected Set<String> sentinels = new HashSet<String>();
```
[the remark this answers is not in the document]
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/SSLJedisSentinelPoolTest.java:24]: @bsbodden I believe he took the idea from `JedisSentinelPoolTest`.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java (hunk 5 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/JedisSentinelPool.java:22]: I'm making this resolved as the PR didn't modify it.

File: src/test/java/redis/clients/jedis/tests/SSLJedisSentinelPoolTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sazzad16) [reply src/test/java/redis/clients/jedis/tests/SSLJedisSentinelPoolTest.java:24]: I'm making this resolved as the code base already has a similar piece of code. Any intended modifications in all of those places can be done in a separate PR.

File: src/main/java/redis/clients/jedis/JedisSentinelPool.java [hunk 7]
```diff
@@ -337,23 +382,39 @@ protected void returnResource(final Jedis resource) {
...
     protected long subscribeRetryWaitTimeMillis = 5000;
     protected volatile Jedis j;
     protected AtomicBoolean running = new AtomicBoolean(false);
 
     protected MasterListener() {
     }
 
-    public MasterListener(String masterName, String host, int port) {
+    public MasterListener(String masterName, String host, int port, boolean isSslEnabled,
+        SSLSocketFactory sslSocketFactory, SSLParameters sslParameters, HostnameVerifier hostnameVerifier) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/JedisSentinelPool.java:397]: This breaks backward compatibility.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**